Model/Product/Variation/Builder.php

```diff
@@ -0,0 +1,151 @@
+use Magento\Customer\Model\Data\Group;
+use Magento\Catalog\Api\Data\ProductTierPriceInterfaceFactory as PriceFactory;
+use Nosto\Object\Product\Product as NostoProduct;
+
+class Builder
+{
+    private $nostoPriceHelper;
+    private $eventManager;
+    private $logger;
+    private $nostoCurrencyHelper;
+    private $priceFactory;
+
+    /**
+     * Builder constructor.
+     *
+     * @param NostoPriceHelper $priceHelper
+     * @param NostoLogger $logger
+     * @param ManagerInterface $eventManager
+     * @param CurrencyHelper $nostoCurrencyHelper
+     * @param PriceFactory $priceFactory
+     */
+    public function __construct(
+        NostoPriceHelper $priceHelper,
+        NostoLogger $logger,
+        ManagerInterface $eventManager,
+        CurrencyHelper $nostoCurrencyHelper,
+        PriceFactory $priceFactory
+    ) {
+        $this->nostoPriceHelper = $priceHelper;
+        $this->logger = $logger;
+        $this->eventManager = $eventManager;
+        $this->nostoCurrencyHelper = $nostoCurrencyHelper;
+        $this->priceFactory = $priceFactory;
+    }
+
+    /**
+     * @param Product $product
+     * @param NostoProduct $
+     * @param $nostoProduct
+     * @param Store $store
+     * @param Group $group
+     * @return Variation
+     */
+    public function build(
+        Product $product,
+        NostoProduct $nostoProduct,
+        Store $store,
+        Group $group
+    ) {
+        $variation = new Variation();
+        try {
+            $variation->setVariationId($group->getCode());
+            $variation->setAvailability($nostoProduct->getAvailability());
+
+            $listPrice = $this->nostoCurrencyHelper->convertToTaggingPrice(
+                $this->nostoPriceHelper->getProductFinalDisplayPrice(
+                    $product,
+                    $store
+                ),
+                $store
+            );
+            $product->setPrice($this->getVariationPrice($product, $group));
+            $price = $this->nostoCurrencyHelper->convertToTaggingPrice(
+                $this->nostoPriceHelper->getProductDisplayPrice(
+                    $product,
+                    $store
+                ),
+                $store
+            );
+
+            $variation->setPrice($price);
+            $variation->setListPrice($listPrice);
+            $variation->setPriceCurrencyCode($nostoProduct->getPriceCurrencyCode());
+        } catch (\Exception $e) {
+            $this->logger->exception($e);
+        }
+
+        $this->eventManager->dispatch(
+            'nosto_variation_load_after',
+            [
+                'variation' => $variation,
+                'magentoProduct' => $product
+            ]
+        );
+        return $variation;
+    }
+
+    /**
+     * @param Product $product
+     * @param Group $group
+     * @return float
+     */
+    private function getVariationPrice(Product $product, Group $group)
+    {
+        $productPrices = $product->getTierPrices();
```

Comment:
Is this work for both the cases including: 

1. Customer Group Price
2, Catalog price rules